### PR TITLE
Fix DefaultExecutor not being able to exit

### PR DESCRIPTION
**Pull request by dkhalanskyjb**
Solves https://github.com/Kotlin/kotlinx.coroutines/issues/856.

### CODE REVIEW

File: kotlinx-coroutines-core/common/src/EventLoop.common.kt [hunk 1]
```diff
@@ -64,11 +67,17 @@ internal abstract class EventLoop : CoroutineDispatcher() {
...
         task.run()
         return true
     }
+
+    public fun popUnconfinedTask(): DispatchedTask<*>? =
+        unconfinedQueue?.removeFirstOrNull()
+
+    protected open fun popNextTask(onlyUnconfined: Boolean = false): Runnable? =
+        popUnconfinedTask()
+
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/EventLoop.common.kt]: Naming nit: `pop` verb is primarily used with stacks (`push`/`pop`), rarely so with queues, especially in JVM & Kotlin ecosystem. I'd suggest using `poll` (ala Java's Queue interface) for a short verb these operations or you can go with long and explicit `removeFirstXxxOrNull`.

File: kotlinx-coroutines-core/common/src/EventLoop.common.kt [hunk 2]
```diff
@@ -249,9 +258,15 @@ internal abstract class EventLoopImplBase: EventLoopImplPlatform(), Delay {
         }
     }
 
-    override fun processNextEvent(): Long {
+    override fun popNextTask(onlyUnconfined: Boolean): Runnable? {
```
Comment (elizarov) [inline kotlinx-coroutines-core/common/src/EventLoop.common.kt]: I cannot find a place where this method if called with `onlyUnconfined = true`.

**Comment by dkhalanskyjb** [comment]
If we do this optimization, then the decomposition of `process` into (poorly named) `pop` and `run` is not needed: I introduced it only so that `DefaultExecutor` could discern between a situation where tasks arrive slower than they are executed (and `process` always returns `Long.MAX_VALUE`) and one where there are no tasks at all (and `process` always returns `Long.MAX_VALUE`).

Should I keep the decomposition or should I revert it and instead implement the small change to the logic of `process` that is also a performance optimization?

**Comment by elizarov** [comment]
@dkhalanskyjb 
> If we do this optimization, then the decomposition of `process` into (poorly named) `pop` and `run` is not needed
It would be also great to avoid decomposition. It will make the code easier to understand.

File: kotlinx-coroutines-core/jvm/test/DefaultExecutorStressTest.kt [hunk 3]
```diff
@@ -35,4 +36,30 @@ class DefaultExecutorStressTest : TestBase() {
         }
         finish(2 + iterations * 4)
     }
+
+    @Test
+    fun workerShutsDown() = withVirtualTimeSource {
```
Comment (elizarov) [inline kotlinx-coroutines-core/jvm/test/DefaultExecutorStressTest.kt]: By convention in this project, test function shall start with the verb `test` like `testWorkerShutsDown`.

Comment (dkhalanskyjb) [reply kotlinx-coroutines-core/jvm/test/DefaultExecutorStressTest.kt]: Yeah, sorry. I was already told that once, so I'll be sure to remember it.

**Pull request merged**